Port WirePattern to Process API

bergie opened the pull request:
Since we have the Process API now, it would solve a lot of WirePattern issues to make it actually just a "configurator" for Process API.

* [x] Make all "normal" WirePattern cases to be handled by Process API
* [x] Mark the more esoteric features (`group`, `field`, `postpone`/`resume` etc) deprecated
* [x] Implement WirePattern-like MultiError handling for the Process API wrapper
* [x] Ensure Process API supports addressable ports correctly #527 
* [x] Implement arrayPolicy in WirePattern Process API wrapper
* [x] Make Process API de-duplicate brackets when forwarding from multiple ports #530
* [x] Implement `group` and `field` collation support with Process API

There are several WirePattern issues to consider:

* Addressable port in WirePattern behaves incorrectly when receiving IP from another component #252
* WirePattern duplicate groups on output #269
* WirePattern doesn't obey arrayPolicy "any" #275
* WirePattern doesn't forward groups correctly when edges to in-port and param-port are both async #276
* Weird doc string in wirepattern #281
* Exception: OUT Socket ID required #293
* Network shutdown may cause WirePattern to fire again #470
* WirePattern adds group on error outports #485
* WirePattern does not respect required: false #490

trustmaster commented:
I think `group` and `field` are not really esoteric. They were one of the original reasons for WirePattern to emerge.

bergie commented:
@trustmaster I know, but they're a little difficult to implement without reading packets from buffer in different order they were added in

trustmaster commented:
@bergie yep, and that's one of the reasons I decided to fix&deprecate WirePattern rather than port it onto Process API back in June.

bergie commented:
OK, with the current state the only failing tests are related to group de-duplication and arrayport handling. Looking pretty promising.

bergie commented:
Now only failing on dupe groups (when we forward `< a < b` from in1 and same `< a <b` from in2 you get `< a < b < a < b`)

jonnor commented:
Why all the changes to the tests in `spec/Helpers.coffee`?

bergie commented:
@jonnor couple of reasons:

* Some tests were reusing same component instance across different WirePattern setups, which doesn't work with Process API
* Cleaned some test interdependencies (using proper before/after etc)
* Added async/sync versions of some tests to verify correctness in both modes
* Some tests were kind of wrong (doing async things in sync wirepattern) that only worked because WirePattern didn't actually care

jonnor commented:
> Some tests were kind of wrong (doing async things in sync wirepattern) that only worked because WirePattern didn't actually care.
There will very likely be components like these in the wild, especially since `async` is opt-in.

In general we're very likely to have components, graphs and programs, which rely on WirePattern "bugs" to function correctly. This makes me very skeptical to forcing use of this new WirePattern implementation. I think we should at least add a way to opt-out, "give me the old WirePattern". Ideally settable for a whole process (for instance via envvar, module-level variable), so one can quickly test whether old/new WirePattern is cause of behavior changes. And per-component, so that one can isolate only the broken component.

bergie commented:
Legacy mode is indeed pretty easy to add since we already have fallback logic in place for couple of the options (`group`, `field`) we didn't add Process API support for yet

bergie commented:
OK, now we should be essentially complete. No fallbacks except if you enable them manually, either with:

* Passing a `legacy: true` to WirePattern
* Setting the `NOFLO_WIREPATTERN_LEGACY` env var

trustmaster commented on spec/Helpers.coffee:
```diff
@@ -643,61 +684,140 @@ describe 'Component traits', ->
         , (data, groups, outs, callback) ->
           setTimeout ->
             if data.num % 2 is 1
-              outs.odd.beginGroup grp for grp in groups
```
Does it mean that `groups` are not captured correctly in this test?

bergie replied:
Yep, `groups` is a bit finicky because in Process API it is resolved properly only when dealing with output queue, so after the processing callback has fired.

bergie replied:
Also, in general this kind of thing being done in this test works better with `forwardGroups: true` than doing it manually

trustmaster commented on spec/Helpers.coffee:
```diff
@@ -1517,22 +1635,23 @@ describe 'Component traits', ->
+          async: true
+          forwardGroups: true
+          group: isUuid
+          gcFrequency: 2 # every 2 requests
```
`gc` options only work in legacy mode i suppose, thus should be removed from here. Also, is `group` by `RegExp` supported?

bergie replied:
grouping by RegExp is supported (top-level group only, see [here](https://github.com/noflo/noflo/pull/526/commits/68c7c52992de1dcf7cda57b41939f69a9d60b675#diff-2d415ed643629dc01235928209dcffc9R464). `gc` is indeed unnecessary here.

trustmaster commented on src/lib/Helpers.coffee:
```diff
@@ -141,40 +124,470 @@ exports.WirePattern = (component, config, proc) ->
+    platform.deprecated 'noflo.helpers.WirePattern field option is deprecated. Please port to Process API'
+  # Then add deprecation warnings for other unwanted behaviors
+  if func.length > 4
+    platform.deprecated 'noflo.helpers.WirePattern postpone and resume are deprecated. Please port to Process API'
```
Are `postpone` and `resume` supported at all? This should be a fatal rather than deprecation warning.

bergie replied:
@trustmaster nope. But the problem is that you might do `(data, groups, out, callback, null, null, scope)` which is why this can't be fatal

bergie replied:
Though I suppose we could check if `postpone` and `resume` are functions and _then_ throw

trustmaster replied:
Yes, that would be much better.

bergie replied:
Well, except we can't see the argument usage from the outside. Plugged the deprecation warning back in. It will throw anyway if user tries to call the (non-provided) function.
